### PR TITLE
server: (UI) add syntax highlighting and latex math rendering

### LINKED ISSUE #10246: web UI : support syntax highlighting
Add support for syntax highlighting in the code fragments of the model output.

### PR DESCRIPTION
**Pull request by VJHack**
*Make sure to read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md) before submitting a PR*

Fixes #10246 and #10758

The changes in this PR include:
- Added support for syntax highlighting in the code fragments of the model output

Screenshots:

Syntax highlighting
image

Light theme:
image

Latex rendering:
image

### CODE REVIEW

File: examples/server/webui/src/main.js
Comment (ngxson) [inline examples/server/webui/src/main.js:11]:
```suggestion
import 'katex/dist/katex.min.css';
```

Please respect existing style

File: examples/server/webui/src/main.js
Comment (ngxson) [inline examples/server/webui/src/main.js:86]:
```suggestion
    md.value.use(HighlightJS, { hljs });
```

Comment (VJHack) [reply examples/server/webui/src/main.js]: Resolved

File: examples/server/webui/src/main.js
Comment (ngxson) [inline examples/server/webui/src/main.js:87]: again, for trivial changes like this, you should pay more attention to the coding style.

it's not fun when your code is just 10 lines and I ended up leaving 3 comments

```suggestion
    md.value.use(markdownItKatexGpt);
```

Comment (VJHack) [reply examples/server/webui/src/main.js]: Thanks for pointing these out. I'll be more careful next time. I appreciate you taking the time to do this. 
Please let me know if there's anything else I can fix or work on. I'm looking for feedback.

It might be a good idea to add a code formatter and linter in the CI or/and pre-commit to prevent this from happening in the future.

**Review by ngxson — APPROVED** [review]
This frontend code is too simple and I don't want to add formatter or linter, it's overkill. Just take 2 or 5 minutes to read the code before committing, is it difficult to do so?

This is an open source project, not a school, so I can't spend there time fixing every possible thing. Because you've already done this way many times before, the next time, I may not review your PR if it takes me too much time. Hope you understand that...

For now, this PR can be merged once the CI passes

**Review by ngxson — CHANGES_REQUESTED** [review]
Hmm one problem, the bundle size gets too big after this change, going from 500kb to more than 2MB now.

I temporarily block merging of this PR, but will see how we can improve that.

**Comment by ngxson** [comment]
katex takes 592KB and highlight.js takes 1.55MB, no tree shaking can be done:

image

**Comment by ngxson** [comment]
I removed math rendering (katex) because it is too big, we can only add it back in the future once we find a way to reduce bundle size

**Review by ngxson — APPROVED** [review]

**Comment by ngxson** [comment]
merging this after https://github.com/ggerganov/llama.cpp/pull/10783 to avoid conflict

**Comment by ngxson** [comment]
Added back katex with gzip to reduce bundle size, as discussed: https://github.com/ggerganov/llama.cpp/issues/10758#issuecomment-2541561836

**Comment by slaren** [comment]
I think this is missing some triggers for latex. Using Qwen2:

image

<details>

`````
Sure! The softmax function can be mathematically formulated as follows:

Given an input vector \(\mathbf{x} = [x_1, x_2, \ldots, x_n]\), the softmax function computes the output vector \(\mathbf{y} = [y_1, y_2, \ldots, y_n]\) where each element \(y_i\) is given by:

\[
y_i = \frac{e^{x_i}}{\sum_{j=1}^{n} e^{x_j}}
\]

This ensures that the output vector \(\mathbf{y}\) is a probability distribution, i.e., all elements are non-negative and sum to 1.

Here is the LaTeX representation of the softmax function:

```latex
y_i = \frac{e^{x_i}}{\sum_{j=1}^{n} e^{x_j}}
```

For the entire vector \(\mathbf{y}\):

```latex
\mathbf{y} = \text{softmax}(\mathbf{x}) = \left[ \frac{e^{x_1}}{\sum_{j=1}^{n} e^{x_j}}, \frac{e^{x_2}}{\sum_{j=1}^{n} e^{x_j}}, \ldots, \frac{e^{x_n}}{\sum_{j=1}^{n} e^{x_j}} \right]
```

This formulation ensures that each element \(y_i\) is a valid probability, and the sum of all elements in \(\mathbf{y}\) is 1.
`````
</detail>

**Comment by ngxson** [comment]
> I think this is missing some triggers for latex. Using Qwen2.5:

Yeah it depends on the model. It must be trained so that it output `\[` to start the latex (instead of `[`). Ref: https://github.com/SchneeHertz/markdown-it-katex-gpt

I'll add a switch in setting for that, so user can switch between `\[` and `[`

**Comment by slaren** [comment]
Note that this output started \[, or with ```latex, and it also failed to render as latex (see the details block for the raw text).

**Comment by ngxson** [comment]
For the one starts with ```latex , it should be an easy fix.

Could you point me to the one starts with \[ ? I can't see those from your screenshot

**Comment by slaren** [comment]
You should have the full raw text in the details block below:
image

**Comment by ngxson** [comment]
This should support pretty much every cases now:

image

````
This is the formula:
$\frac{e^{x_i}}{\sum_{j=1}^{n}e^{x_j}}$

Given an input vector \(\mathbf{x} = [x_1, x_2, \ldots, x_n]\)

\[
y_i = \frac{e^{x_i}}{\sum_{j=1}^n e^{x_j}}
\]

Code block latex:
```latex
\frac{e^{x_i}}{\sum_{j=1}^{n}e^{x_j}}
```

Test dollar sign: $1234 $4567

Invalid latex syntax: $E = mc^$ and $$E = mc^$$
````

**Comment by slaren** [comment]
Looking back, I think rendering the latex code block as latex is a mistake, it should be just latex code with syntax highlight if available. I asked the model to show me softmax with latex, and that's what it did.

**Comment by ngxson** [comment]
yeah that makes sense, as user may ask the chatbot "how to write latex" for example. the code should be shown as-it in codeblock in this case.

also, in the future, we can have a switch to enabled/disable rendering latex in codeblock if someone need that

**Review by slaren — APPROVED** [review]

**Comment by ngxson** [comment]
@ggerganov I'm gonna merge this PR now. This will produce a conflict when you merge your other PR.

To resolve it, you just need to delete `examples/server/public/index.html` (because it's now a `.gz` file) and re-run `npm run build`, it will re-generate `index.html.gz`

**Pull request merged**
